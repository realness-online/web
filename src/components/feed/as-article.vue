<template lang="html">
  <article class="feed post">
    <header>
      <router-link :to="person.id">
        <profile-as-avatar :person="person"></profile-as-avatar>
      </router-link>
      <hgroup>
        <span>{{person.first_name}} {{person.last_name}}</span>
        <time :datetime="post.created_at">{{as_created_time}}</time>
      </hgroup>
    </header>
    <p>{{as_statement}}</p>
    <p v-for="statement in post.statements" :key="statement.id" >{{as_statement_from_post(statement)}}</p>
  </article>
</template>
<script>
  import post_mixin from '@/mixins/post'
  import date_mixin from '@/mixins/date'
  import intersection_mixin from '@/mixins/intersection'
  import profile_as_avatar from '@/components/avatars/as-svg'
  export default {
    mixins: [post_mixin, date_mixin, intersection_mixin],
    components: {
      'profile-as-avatar': profile_as_avatar
    }
  }
</script>
<style lang="stylus">
  article.feed.post
    overflow: hidden
    margin-bottom: base-line
    &:last-of-type
      margin-bottom: 0
</style>
