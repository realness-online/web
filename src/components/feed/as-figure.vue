<template lang="html">
  <figure class='feed poster'>
    <figcaption>
      <router-link :to="poster.person.id">
        <profile-as-avatar :person="poster.person"></profile-as-avatar>
      </router-link>
      <hgroup>
        <span class="person">{{poster.person.first_name}} {{poster.person.last_name}}</span>
        <time :datetime="poster.created_at">{{as_time(poster.created_at)}}</time>
      </hgroup>
    </figcaption>
    <icon v-if="actual_poster" name="background"></icon>
    <svg v-if="actual_poster" @click="toggle_slice" :preserveAspectRatio="aspect_ratio"
         :viewBox="actual_poster.view_box" v-html="actual_poster.path">
    </svg>
    <icon v-else name="working"></icon>
    <menu v-if="!slice">
      <download-vector :vector="actual_poster" :author="poster.person"></download-vector>
    </menu>
  </figure>
</template>
<script>
  import profile from '@/helpers/profile'
  import date_mixin from '@/mixins/date'
  import icon from '@/components/icon'
  import download_vector from '@/components/download-vector'
  import profile_as_avatar from '@/components/avatars/as-svg'
  export default {
    mixins: [date_mixin],
    props: ['poster'],
    components: {
      'profile-as-avatar': profile_as_avatar,
      'download-vector': download_vector,
      icon
    },
    data() {
      return {
        slice: true,
        observer: null,
        actual_poster: null,
        options: {
          rootMargin: '0px 0px 0px 0px'
        }
      }
    },
    async mounted() {
      this.observer = new IntersectionObserver(this.show_poster, this.options)
      await this.$nextTick()
      this.observer.observe(this.$el)
    },
    destroyed() {
      if (this.observer) this.observer.unobserve(this.$el)
    },
    computed: {
      aspect_ratio() {
        if (this.slice) return 'xMidYMid slice'
        else return 'xMidYMid meet'
      }
    },
    methods: {
      toggle_slice() {
        this.slice = !this.slice
      },
      async show_poster(entries) {
        entries.forEach(async entry => {
          if (entry.isIntersecting) {
            this.actual_poster = await profile.item(this.poster.person.id, this.poster.id)
            this.observer.unobserve(this.$el)
          }
        })
      }
    }
  }
</script>
<style lang="stylus">
  figure.feed.poster
    position:relative
    & >  menu
      display: flex
      justify-content:flex-end
      padding: base-line
      margin-top: -(base-line * 4)
      & > a > svg
        fill: red
    & > svg
      width: stretch
      height: 100vh
      max-height: page-width + (base-line * 6)
      &.background
        fill: blue
      &.working
        height:auto
        width: base-line * 5
      @media (max-width: min-screen)
        &:not(.working)
          margin-left -(base-line)
          margin-right -(base-line)
</style>
