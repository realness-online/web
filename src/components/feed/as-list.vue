<template lang="html">
  <div v-for="[page_name, days] in pages" :key="page_name">
    <section class="day" v-for="[date, day] in days" :key="date" v-bind:class="{today: is_today(date)}">
      <header><h4>{{as_day(date)}}</h4></header>
      <div v-for="post in day" :key="post.id">
        <poster-as-figure v-if="post.type === 'posters'"
                          :poster="post">
        </poster-as-figure>
        <post-as-article v-else
                         :post="post"
                         :person="post.person"
                         @end-of-articles="next_page">
        </post-as-article>
      </div>
    </section>
</template>
<script>
  import post_as_article from '@/components/feed/as-article'
  import poster_as_figure from '@/components/feed/as-figure'
  export default {
    components: {
      'post-as-article': post_as_article,
      'poster-as-figure': poster_as_figure
    },
    methods: {
      async get_first_posts(people_in_feed) {
        let feed = []
        await Promise.all(people_in_feed.map(async (relation) => {
          const [person, posts, posters] = await Promise.all([
            profile.load(relation.id),
            profile.items(relation.id, 'posts/index'),
            profile.directory(relation.id, 'posters')
          ])
          this.relations.push(person)
          feed = [...this.condense_posts(posts, person),
                  ...this.prepare_posters(posters, person), ...feed]
        }))
        feed.sort(this.newer_first)
        feed.forEach(post => this.insert_post_into_day(post, this.days))
      },
      async next_page(person) {
        if (person.page) person.page = growth.next(person.page)
        else person.page = growth.first()
        console.log(`posts.${person.page}`, person.first_name)
        let posts = await profile.items(person.id, `posts/${person.page}`)
        posts = this.condense_posts(posts, person)
        posts.forEach(post => this.insert_post_into_day(post, this.days))
        const sorted = [...this.days.entries()].sort(this.newer_day_first)
        this.days = new Map(sorted)
      }
    }
  }
</script>
<style lang="stylus">
</style>
