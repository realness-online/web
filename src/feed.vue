<template lang="html">
  <section id="feed">

    <details>
      <summary><h3>Feed</h3></summary>
      <p>
        So sad.
        <a href="/groups"> So very very sad</a>
      </p>
    </details>
    <details>
      <summary><h3>Activity</h3> </summary>
      <activity-list :activity="activity"></activity-list>
    </details>
    <details>
      <summary><h3>Posts</h3></summary>
      <posts-list :posts="posts"></posts-list>
    </details>
  </section>
</template>

<style lang='stylus'>
  @require 'application'
  section#feed
    display:flex
    flex-direction: column
    padding: 0 base-line
    & > details[open]
      min-height:100vh
</style>

<script>
  import {posts_storage, activity_storage} from '@/modules/Storage'
  import posts_list from '@/components/posts-list'
  import activity_list from '@/components/activity-list'
  import '@/modules/timeago'
  export default {
    components: {
      'activity-list': activity_list,
      'posts-list': posts_list
    },
    data() {
      return {
        activity: activity_storage.get_items(),
        posts: posts_storage.get_items()
      }
    }
  }
</script>
