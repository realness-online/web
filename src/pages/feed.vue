<template>
  <section id="feed" class="page left">
    <header>
      <icon name="hamburger"></icon>
      <h1>Feed</h1>
      <logo-as-link></logo-as-link>
    </header>
    <icon v-show="working" name="working"></icon>
    <article v-for="post in feed" itemscope itemtype="/post">
      <header>
        <profile-as-figure :person='post.person'></profile-as-figure>
      </header>
      <blockquote itemprop="articleBody">{{post.articleBody}}</blockquote>
      <time itemprop="created_at" :datetime="post.created_at">calculating...</time>
    </article>
  </section>
</template>
<script>
  import * as firebase from 'firebase/app'
  import 'firebase/storage'
  import Storage, {relations_storage} from '@/modules/Storage'
  import Item from '@/modules/Item'
  import logoAsLink from '@/components/logo-as-link'
  import profileAsFigure from '@/components/profile/as-figure'
  import icon from '@/components/icon'
  export default {
    components: {
      profileAsFigure,
      logoAsLink,
      icon
    },
    data() {
      return {
        feed: [],
        working: true
      }
    },
    methods: {
      posts_url(person) {
        let posts_path = `/people/+1${person.mobile}/posts.html`
        return firebase.storage().ref().child(posts_path).getDownloadURL()
      }
    },
    created() {
      relations_storage.as_list().forEach((person) => {
        this.posts_url(person).then((url) => {
          fetch(url).then((response) => {
            response.text().then(text => {
              let posts = Item.get_items(Storage.hydrate(text))
              posts.forEach(post => (post.person = person))
              this.feed.push(...posts)
              this.feed.sort((a, b) => {
                return Date.parse(b.created_at) - Date.parse(a.created_at)
              })
              this.working = false
            })
          })
        })
      })
    }
  }
  // check if their updated since last time.
  // if their are new updates then update Feed
  // this requires that posts update the updated at property for each person
  // when they create a post
  // let profile_path = `/people/+1${person.mobile}/profile.html`
  // this.get_item_by_path(relations_path).then((network_person) => {
  //   if (network_person.updated_at > person.updated_at){
  //
  //     console.log('getting updates for:', person.first_name)
  //
  //   }
</script>
<style lang="stylus">
  @require '../style/variables'
  section#feed
    display: flex
    flex-direction: column-reverse
    & > header
      order: 2
      margin-bottom: base-line
      & > svg
        fill: transparent
    & > svg.working
      order: 1
      margin-bottom: base-line
    & > article
      margin-bottom: base-line
      & > header
        margin-bottom: (base-line / 2)
        & > figure a svg
          fill: blue
          stroke: lighten(blue, 33%)
          border-radius: base-line
      & > blockquote
        margin-bottom: (base-line / 2)
</style>
