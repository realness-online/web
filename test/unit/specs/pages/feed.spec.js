import Vue from 'vue'
import { shallow } from 'vue-test-utils'
import feed from '@/pages/feed'

describe('@/pages/feed.vue', () => {
  it('render feed info', () => {
    let wrapper = shallow(feed)
    expect(wrapper.element).toMatchSnapshot()
  })
})
