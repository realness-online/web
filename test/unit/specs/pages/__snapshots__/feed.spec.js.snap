// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/pages/feed.vue render feed info 1`] = `
<section
  class="page"
  id="feed"
>
  <header>
    <!---->
     
    <h1>
      Feed
    </h1>
     
    <!---->
  </header>
</section>
`;
